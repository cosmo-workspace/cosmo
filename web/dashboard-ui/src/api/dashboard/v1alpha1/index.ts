/* tslint:disable */
/* eslint-disable */
/**
 * Cosmo Dashboard API
 * Manipulate cosmo dashboard resource API
 *
 * The version of the OpenAPI document: v1alpha1
 * Contact: jlandowner8@gmail.com
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */


export * from "./api";
export * from "./configuration";

