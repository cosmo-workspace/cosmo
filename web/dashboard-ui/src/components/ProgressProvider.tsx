import { Backdrop, CircularProgress } from '@mui/material';
import React, { createContext, useContext, useState } from 'react';

const DispatchContext = createContext<React.Dispatch<React.SetStateAction<number>>>(undefined as any);

/**
 * provider
 */
export const ProgressProvider: React.FC = ({ children }) => {
  const [count, setCount] = useState(0);
  return (<>
    <DispatchContext.Provider value={setCount}>
      {children}
    </DispatchContext.Provider>
    <Backdrop sx={{ zIndex: (theme) => theme.zIndex.drawer + 1000 }} open={count > 0}>
      <CircularProgress />
    </Backdrop>
  </>);
}

export function useProgress() {
  const setCount = useContext(DispatchContext);
  return {
    setMask: () => setCount(count => { console.log('setMask', count + 1); return count + 1 }),
    releaseMask: () => setCount(count => { console.log('releaseMask', count); return count > 0 ? count - 1 : 0 }),
  }
}