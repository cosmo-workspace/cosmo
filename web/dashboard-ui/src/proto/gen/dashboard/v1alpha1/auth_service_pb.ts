//
//Cosmo Dashboard API
//Manipulate cosmo dashboard resource API

// @generated by protoc-gen-es v1.2.0 with parameter "target=ts"
// @generated from file dashboard/v1alpha1/auth_service.proto (package dashboard.v1alpha1, syntax proto3)
/* eslint-disable */
// @ts-nocheck

import type { BinaryReadOptions, FieldList, JsonReadOptions, JsonValue, PartialMessage, PlainMessage } from "@bufbuild/protobuf";
import { Message, proto3, Timestamp } from "@bufbuild/protobuf";

/**
 * @generated from message dashboard.v1alpha1.LoginRequest
 */
export class LoginRequest extends Message<LoginRequest> {
  /**
   * @generated from field: string user_name = 1;
   */
  userName = "";

  /**
   * @generated from field: string password = 2;
   */
  password = "";

  constructor(data?: PartialMessage<LoginRequest>) {
    super();
    proto3.util.initPartial(data, this);
  }

  static readonly runtime: typeof proto3 = proto3;
  static readonly typeName = "dashboard.v1alpha1.LoginRequest";
  static readonly fields: FieldList = proto3.util.newFieldList(() => [
    { no: 1, name: "user_name", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "password", kind: "scalar", T: 9 /* ScalarType.STRING */ },
  ]);

  static fromBinary(bytes: Uint8Array, options?: Partial<BinaryReadOptions>): LoginRequest {
    return new LoginRequest().fromBinary(bytes, options);
  }

  static fromJson(jsonValue: JsonValue, options?: Partial<JsonReadOptions>): LoginRequest {
    return new LoginRequest().fromJson(jsonValue, options);
  }

  static fromJsonString(jsonString: string, options?: Partial<JsonReadOptions>): LoginRequest {
    return new LoginRequest().fromJsonString(jsonString, options);
  }

  static equals(a: LoginRequest | PlainMessage<LoginRequest> | undefined, b: LoginRequest | PlainMessage<LoginRequest> | undefined): boolean {
    return proto3.util.equals(LoginRequest, a, b);
  }
}

/**
 * @generated from message dashboard.v1alpha1.LoginResponse
 */
export class LoginResponse extends Message<LoginResponse> {
  /**
   * @generated from field: string user_name = 1;
   */
  userName = "";

  /**
   * @generated from field: google.protobuf.Timestamp expire_at = 2;
   */
  expireAt?: Timestamp;

  /**
   * @generated from field: bool require_password_update = 3;
   */
  requirePasswordUpdate = false;

  constructor(data?: PartialMessage<LoginResponse>) {
    super();
    proto3.util.initPartial(data, this);
  }

  static readonly runtime: typeof proto3 = proto3;
  static readonly typeName = "dashboard.v1alpha1.LoginResponse";
  static readonly fields: FieldList = proto3.util.newFieldList(() => [
    { no: 1, name: "user_name", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "expire_at", kind: "message", T: Timestamp },
    { no: 3, name: "require_password_update", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
  ]);

  static fromBinary(bytes: Uint8Array, options?: Partial<BinaryReadOptions>): LoginResponse {
    return new LoginResponse().fromBinary(bytes, options);
  }

  static fromJson(jsonValue: JsonValue, options?: Partial<JsonReadOptions>): LoginResponse {
    return new LoginResponse().fromJson(jsonValue, options);
  }

  static fromJsonString(jsonString: string, options?: Partial<JsonReadOptions>): LoginResponse {
    return new LoginResponse().fromJsonString(jsonString, options);
  }

  static equals(a: LoginResponse | PlainMessage<LoginResponse> | undefined, b: LoginResponse | PlainMessage<LoginResponse> | undefined): boolean {
    return proto3.util.equals(LoginResponse, a, b);
  }
}

/**
 * @generated from message dashboard.v1alpha1.VerifyResponse
 */
export class VerifyResponse extends Message<VerifyResponse> {
  /**
   * @generated from field: string user_name = 1;
   */
  userName = "";

  /**
   * @generated from field: google.protobuf.Timestamp expire_at = 2;
   */
  expireAt?: Timestamp;

  /**
   * @generated from field: bool require_password_update = 3;
   */
  requirePasswordUpdate = false;

  constructor(data?: PartialMessage<VerifyResponse>) {
    super();
    proto3.util.initPartial(data, this);
  }

  static readonly runtime: typeof proto3 = proto3;
  static readonly typeName = "dashboard.v1alpha1.VerifyResponse";
  static readonly fields: FieldList = proto3.util.newFieldList(() => [
    { no: 1, name: "user_name", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "expire_at", kind: "message", T: Timestamp },
    { no: 3, name: "require_password_update", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
  ]);

  static fromBinary(bytes: Uint8Array, options?: Partial<BinaryReadOptions>): VerifyResponse {
    return new VerifyResponse().fromBinary(bytes, options);
  }

  static fromJson(jsonValue: JsonValue, options?: Partial<JsonReadOptions>): VerifyResponse {
    return new VerifyResponse().fromJson(jsonValue, options);
  }

  static fromJsonString(jsonString: string, options?: Partial<JsonReadOptions>): VerifyResponse {
    return new VerifyResponse().fromJsonString(jsonString, options);
  }

  static equals(a: VerifyResponse | PlainMessage<VerifyResponse> | undefined, b: VerifyResponse | PlainMessage<VerifyResponse> | undefined): boolean {
    return proto3.util.equals(VerifyResponse, a, b);
  }
}

