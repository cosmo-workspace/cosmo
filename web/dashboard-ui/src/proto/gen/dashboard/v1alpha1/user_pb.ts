//
//Cosmo Dashboard API
//Manipulate cosmo dashboard resource API

// @generated by protoc-gen-es v1.2.0 with parameter "target=ts"
// @generated from file dashboard/v1alpha1/user.proto (package dashboard.v1alpha1, syntax proto3)
/* eslint-disable */
// @ts-nocheck

import type { BinaryReadOptions, FieldList, JsonReadOptions, JsonValue, PartialMessage, PlainMessage } from "@bufbuild/protobuf";
import { Message, proto3 } from "@bufbuild/protobuf";

/**
 * @generated from message dashboard.v1alpha1.User
 */
export class User extends Message<User> {
  /**
   * @generated from field: string name = 1;
   */
  name = "";

  /**
   * @generated from field: string display_name = 2;
   */
  displayName = "";

  /**
   * @generated from field: repeated string roles = 3;
   */
  roles: string[] = [];

  /**
   * @generated from field: string auth_type = 4;
   */
  authType = "";

  /**
   * @generated from field: repeated dashboard.v1alpha1.UserAddons addons = 5;
   */
  addons: UserAddons[] = [];

  /**
   * @generated from field: string default_password = 6;
   */
  defaultPassword = "";

  /**
   * @generated from field: string status = 7;
   */
  status = "";

  constructor(data?: PartialMessage<User>) {
    super();
    proto3.util.initPartial(data, this);
  }

  static readonly runtime: typeof proto3 = proto3;
  static readonly typeName = "dashboard.v1alpha1.User";
  static readonly fields: FieldList = proto3.util.newFieldList(() => [
    { no: 1, name: "name", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "display_name", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 3, name: "roles", kind: "scalar", T: 9 /* ScalarType.STRING */, repeated: true },
    { no: 4, name: "auth_type", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 5, name: "addons", kind: "message", T: UserAddons, repeated: true },
    { no: 6, name: "default_password", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 7, name: "status", kind: "scalar", T: 9 /* ScalarType.STRING */ },
  ]);

  static fromBinary(bytes: Uint8Array, options?: Partial<BinaryReadOptions>): User {
    return new User().fromBinary(bytes, options);
  }

  static fromJson(jsonValue: JsonValue, options?: Partial<JsonReadOptions>): User {
    return new User().fromJson(jsonValue, options);
  }

  static fromJsonString(jsonString: string, options?: Partial<JsonReadOptions>): User {
    return new User().fromJsonString(jsonString, options);
  }

  static equals(a: User | PlainMessage<User> | undefined, b: User | PlainMessage<User> | undefined): boolean {
    return proto3.util.equals(User, a, b);
  }
}

/**
 * @generated from message dashboard.v1alpha1.UserAddons
 */
export class UserAddons extends Message<UserAddons> {
  /**
   * @generated from field: string template = 1;
   */
  template = "";

  /**
   * @generated from field: bool cluster_scoped = 2;
   */
  clusterScoped = false;

  /**
   * @generated from field: map<string, string> vars = 3;
   */
  vars: { [key: string]: string } = {};

  constructor(data?: PartialMessage<UserAddons>) {
    super();
    proto3.util.initPartial(data, this);
  }

  static readonly runtime: typeof proto3 = proto3;
  static readonly typeName = "dashboard.v1alpha1.UserAddons";
  static readonly fields: FieldList = proto3.util.newFieldList(() => [
    { no: 1, name: "template", kind: "scalar", T: 9 /* ScalarType.STRING */ },
    { no: 2, name: "cluster_scoped", kind: "scalar", T: 8 /* ScalarType.BOOL */ },
    { no: 3, name: "vars", kind: "map", K: 9 /* ScalarType.STRING */, V: {kind: "scalar", T: 9 /* ScalarType.STRING */} },
  ]);

  static fromBinary(bytes: Uint8Array, options?: Partial<BinaryReadOptions>): UserAddons {
    return new UserAddons().fromBinary(bytes, options);
  }

  static fromJson(jsonValue: JsonValue, options?: Partial<JsonReadOptions>): UserAddons {
    return new UserAddons().fromJson(jsonValue, options);
  }

  static fromJsonString(jsonString: string, options?: Partial<JsonReadOptions>): UserAddons {
    return new UserAddons().fromJsonString(jsonString, options);
  }

  static equals(a: UserAddons | PlainMessage<UserAddons> | undefined, b: UserAddons | PlainMessage<UserAddons> | undefined): boolean {
    return proto3.util.equals(UserAddons, a, b);
  }
}

