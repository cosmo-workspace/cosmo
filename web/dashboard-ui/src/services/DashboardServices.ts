import { createPromiseClient } from "@bufbuild/connect";
import { createConnectTransport, createGrpcWebTransport } from "@bufbuild/connect-web";
import { useMemo } from "react";
import { AuthService } from "../proto/gen/dashboard/v1alpha1/auth_service_connectweb";
import { TemplateService } from "../proto/gen/dashboard/v1alpha1/template_service_connectweb";
import { UserService } from "../proto/gen/dashboard/v1alpha1/user_service_connectweb";
import { WorkspaceService } from "../proto/gen/dashboard/v1alpha1/workspace_service_connectweb";

const transportX = createConnectTransport({
    baseUrl: import.meta.env.BASE_URL,
});

const transport = createGrpcWebTransport({
    baseUrl: import.meta.env.BASE_URL,
});

export function useAuthService() {
    return useMemo(() => createPromiseClient(AuthService, transport), [AuthService]);
}
export function useTemplateService() {
    return useMemo(() => createPromiseClient(TemplateService, transport), [TemplateService]);
}
export function useUserService() {
    return useMemo(() => createPromiseClient(UserService, transport), [UserService]);
}

export function useWorkspaceService() {
    return useMemo(() => createPromiseClient(WorkspaceService, transport), [WorkspaceService]);
}