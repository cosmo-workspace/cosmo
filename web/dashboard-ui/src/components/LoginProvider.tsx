import { Box, CircularProgress, CssBaseline, Stack, Typography } from '@mui/material';
import { useSnackbar } from 'notistack';
import React, { createContext, useContext, useEffect, useState } from 'react';
import { User } from '../proto/gen/dashboard/v1alpha1/user_pb';
import { useAuthService, useUserService } from '../services/DashboardServices';
import { useProgress } from './ProgressProvider';


/**
 * context
 * ref: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/24509#issuecomment-774430643
 */
const Context = createContext<ReturnType<typeof useLoginModule>>(undefined as any);

/**
 * hooks
 */
export function useLogin() {
  return useContext(Context);
}

const useLoginModule = () => {
  console.log('useLoginModule');
  const [loginUser, setLoginUser] = useState<User>();
  const { enqueueSnackbar } = useSnackbar();
  const { setMask, releaseMask } = useProgress();
  const userService = useUserService();
  const authService = useAuthService();


  /**
   * verifyLogin
   */
  const verifyLogin = async () => {
    console.log('verify start');
    try {
      const resp = await authService.verify({});
      if (resp.userName) {
        await getMyUserInfo(resp.userName);
      }
    }
    catch (error) {
      setLoginUser(undefined);
      //handleError(error);
    }
    finally {
      console.log('verify end');
    }
  }


  /**
   * login: SignIn 
   */
  const login = async (userName: string, password: string) => {
    console.log('login');
    try {
      const res = await authService.login({ userName: userName, password: password });
      await getMyUserInfo(userName);
      console.log('login end');
      return res;
    }
    catch (error) {
      setLoginUser(undefined);
      console.log('login error');
      handleError(error);
      throw error;
    }
  }


  /**
   * login: MyUserInfo
   */
  const getMyUserInfo = async (userName: string) => {
    console.log('getMyUserInfo', userName);
    // if (loginUser || !username) {
    //   console.log('getMyUserInfo cancel', loginUser, username);
    //   return;
    // }
    try {
      const responseUser = await userService.getUser({ userName: userName });
      setLoginUser(prev => { console.log('setLoginUser', prev, responseUser.user); return responseUser.user! });
    }
    catch (error) {
      console.log('getMyUserInfo error');
      handleError(error);
      throw error;
    }
  }

  /**
   * login: MyUserInfo
   */
  const refreshUserInfo = async () => {
    console.log('refreshUserInfo');
    if (loginUser) {
      getMyUserInfo(loginUser.name);
    }
  }

  /**
   * logout: Pagetemplate 
   */
  const logout = async () => {
    console.log('logout');
    try {
      await authService.logout({});
    }
    catch (error) {
      console.log('logout error');
      handleError(error);
      throw error;
    }
    finally {
      console.log('logout end');
      setLoginUser(prev => { console.log('setLoginUser', `${prev} -> undefined`); return undefined });
    }
  }

  /**
   * updataPassword
   */
  const updataPassword = async (currentPassword: string, newPassword: string) => {
    console.log('updataPassword', loginUser?.name);
    setMask();
    try {
      try {
        return await userService.updateUserPassword({ userName: loginUser!.name, currentPassword, newPassword });
      }
      catch (error) {
        handleError(error);
        throw error;
      }
    }
    finally {
      console.log('updataPassword end');
      releaseMask();
    }
  }

  /**
   * clearLoginUser 
   */
  const clearLoginUser = async () => {
    setLoginUser(undefined);
  }

  /**
   * error handler
   */
  const handleError = (error: any) => {
    console.log(error);
    const msg = error?.message;
    msg && enqueueSnackbar(msg, { variant: 'error' });
  }

  return {
    loginUser,
    verifyLogin,
    login,
    logout,
    refreshUserInfo,
    updataPassword,
    clearLoginUser,
  };
}

/**
 * Provider
 */
export const LoginProvider: React.FC<React.PropsWithChildren<unknown>> = ({ children }) => {
  console.log('LoginProvider');
  const loginModule = useLoginModule();
  const [isVerified, setIsVerified] = useState(false);

  useEffect(() => {
    loginModule.verifyLogin()
      .then(() => setIsVerified(true));
  }, []); // eslint-disable-line

  return (
    <Context.Provider value={loginModule}>
      {isVerified ? children :
        <div>
          <CssBaseline />
          <Stack sx={{ m: 10 }} alignItems='center' spacing={2}>
            <Box sx={{ display: 'flex' }}>
              <CircularProgress />
            </Box>
            <Typography >verifying session...</Typography>
          </Stack>
        </div>}
    </Context.Provider >
  )
}

/**
 * export private member. (for test)
 */
// export const __local__ = {
//   useLoginModule,
// };
