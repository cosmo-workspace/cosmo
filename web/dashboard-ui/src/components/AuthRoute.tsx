import React, { ReactElement } from 'react';
import { Navigate, useLocation } from 'react-router-dom';
import { isAdminUser } from '../views/organisms/UserModule';
import { useLogin } from './LoginProvider';

type Props = {
  children: ReactElement;
  admin?: boolean;
}

export const AuthRoute: React.VFC<Props> = ({ children, admin }) => {
  const { loginUser } = useLogin();
  const isAdmin = isAdminUser(loginUser);
  let location = useLocation();

  if (!loginUser) {
    return (<Navigate
      to={{ pathname: '/signin', search: `from=${location.pathname}` }}
      state={{ from: location }}
      replace
    />);
  } else if (admin && !isAdmin) {
    return (<Navigate to='/' replace />);
  } else {
    return children;
  }
}