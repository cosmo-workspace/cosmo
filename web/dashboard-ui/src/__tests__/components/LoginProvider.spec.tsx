import { Timestamp } from "@bufbuild/protobuf";
import { cleanup, renderHook, waitFor } from "@testing-library/react";
import { useSnackbar } from "notistack";
import React from 'react';
import { act } from "react-dom/test-utils";
import { afterEach, beforeEach, describe, expect, it, MockedFunction, vi } from "vitest";
import { LoginProvider, useLogin } from "../../components/LoginProvider";
import { useProgress } from "../../components/ProgressProvider";
import { LoginResponse, VerifyResponse } from "../../proto/gen/dashboard/v1alpha1/auth_service_pb";
import { User } from "../../proto/gen/dashboard/v1alpha1/user_pb";
import { GetUserResponse, UpdateUserPasswordResponse } from "../../proto/gen/dashboard/v1alpha1/user_service_pb";
import { useAuthService, useUserService } from "../../services/DashboardServices";

//--------------------------------------------------
// mock definition
//--------------------------------------------------

vi.mock('notistack');
vi.mock('../../services/DashboardServices');
vi.mock('../../components/ProgressProvider');
vi.mock('react-router-dom', () => ({
  //useHistory: vi.fn(),
}));


type MockedMemberFunction<T extends (...args: any) => any> = {
  [P in keyof ReturnType<T>]: MockedFunction<ReturnType<T>[P]>;
};

const useSnackbarMock = useSnackbar as MockedFunction<typeof useSnackbar>;
const snackbarMock: MockedMemberFunction<typeof useSnackbar> = {
  enqueueSnackbar: vi.fn(),
  closeSnackbar: vi.fn(),
}

const useProgressMock = useProgress as MockedFunction<typeof useProgress>;
const progressMock: MockedMemberFunction<typeof useProgress> = {
  setMask: vi.fn(),
  releaseMask: vi.fn(),
}

const authService = useAuthService as MockedFunction<typeof useAuthService>;
const authMock: MockedMemberFunction<typeof useAuthService> = {
  verify: vi.fn(),
  login: vi.fn(),
  logout: vi.fn(),
}

const userService = useUserService as MockedFunction<typeof useUserService>;
const userMock: MockedMemberFunction<typeof useUserService> = {
  getUser: vi.fn(),
  getUsers: vi.fn(),
  deleteUser: vi.fn(),
  createUser: vi.fn(),
  updateUserDisplayName: vi.fn(),
  updateUserPassword: vi.fn(),
  updateUserRole: vi.fn()
}

//-----------------------------------------------
// test
//-----------------------------------------------

describe('LoginProvider', () => {

  beforeEach(async () => {
    useSnackbarMock.mockReturnValue(snackbarMock);
    useProgressMock.mockReturnValue(progressMock);
    authService.mockReturnValue(authMock);
    userService.mockReturnValue(userMock);
  });

  afterEach(() => {
    // expect(snackbarMock.enqueueSnackbar.mock.calls).toMatchSnapshot('enqueueSnackbar calls');
    expect(snackbarMock.enqueueSnackbar.mock.calls).toMatchSnapshot();
    vi.restoreAllMocks();
    cleanup();
  });

  async function renderUseLogin() {
    const utils = renderHook(() => useLogin(), {
      wrapper: ({ children }) => (<LoginProvider>{children}</LoginProvider >),
    });
    await waitFor(async () => { expect(utils.result.current).not.toBeNull(); });
    return utils;
  }

  describe('verify', () => {

    it('✅ ok', async () => {
      authMock.verify.mockResolvedValueOnce(new VerifyResponse({
        userName: 'user1',
        expireAt: Timestamp.fromDate(new Date("2022/11/4")),
        requirePasswordUpdate: false,
      }));
      userMock.getUser.mockResolvedValue(new GetUserResponse({
        user: new User({ name: 'user1', roles: ["CosmoAdmin"], displayName: 'user1 name' }),
      }));

      const { result } = await renderUseLogin();
      expect(result.current.loginUser).toMatchSnapshot();
    });


    it('❌ ng / id is undefined', async () => {
      authMock.verify.mockResolvedValueOnce(new VerifyResponse({
        userName: undefined as any,
        expireAt: Timestamp.fromDate(new Date("2022/11/4")),
        requirePasswordUpdate: false,
      }));

      const { result } = await renderUseLogin();
      expect(result.current.loginUser).toMatchSnapshot();
    });


    it('❌ ng', async () => {

      authMock.verify.mockRejectedValue(new Error('[mock] verify error'));
      const { result } = await renderUseLogin();
      expect(result.current.loginUser).toMatchSnapshot();
    });

    it('❌ getUser error', async () => {

      authMock.verify.mockResolvedValueOnce(new VerifyResponse({
        userName: 'user1',
        expireAt: Timestamp.fromDate(new Date("2022/11/4")),
        requirePasswordUpdate: false,
      }));

      userMock.getUser.mockRejectedValue(new Error('[mock] getUser error'));

      const { result } = await renderUseLogin();
      expect(result.current.loginUser).toMatchSnapshot();
    });

  });


  describe('login', () => {

    it('✅ ok', async () => {
      authMock.login.mockResolvedValue(new LoginResponse({
        userName: 'user1',
        expireAt: Timestamp.fromDate(new Date("2022/11/4")),
        requirePasswordUpdate: true,
      }));
      userMock.getUser.mockResolvedValue(new GetUserResponse({
        user: new User({ name: 'user1', roles: ["CosmoAdmin"], displayName: 'user1 name' }),
      }));
      const { result } = await renderUseLogin();
      // await act(async () => {
      await expect(result.current.login('user1', 'password1')).resolves.toMatchSnapshot();
      // });
      await waitFor(async () => { expect(result.current.loginUser).not.toBeUndefined(); });
      expect(result.current.loginUser).toMatchSnapshot();
    });


    it('❌ ng', async () => {
      authMock.login.mockResolvedValue(new LoginResponse({
        userName: 'user1',
        expireAt: Timestamp.fromDate(new Date("2022/11/4")),
        requirePasswordUpdate: true,
      }));
      userMock.getUser.mockResolvedValue(new GetUserResponse({
        user: { name: 'user1', roles: ["CosmoAdmin"], displayName: 'user1 name' },
      }));
      authMock.login.mockRejectedValue(new Error('[mock] login error'));
      const { result } = await renderUseLogin();
      await act(async () => {
        await expect(result.current.login('user1', 'password1')).rejects.toMatchSnapshot();
      });
      await waitFor(async () => { expect(result.current).not.toBeNull(); });
      expect(result.current.loginUser).toMatchSnapshot();
    });


    it('❌ getUser error', async () => {
      const { result } = await renderUseLogin();

      authMock.login.mockResolvedValue(new LoginResponse({
        userName: 'user1',
        expireAt: Timestamp.fromDate(new Date("2022/11/4")),
        requirePasswordUpdate: true,
      }));

      userMock.getUser.mockRejectedValue(new Error('[mock] getUser error'));

      await act(async () => {
        await expect(result.current.login('user1', 'password1')).rejects.toMatchSnapshot();
      });
      expect(result.current.loginUser).toMatchSnapshot();
    });

  });


  describe('refreshUserInfo', () => {

    describe('not login', () => {

      it('✅ ok', async () => {
        const { result } = await renderUseLogin();
        await expect(result.current.refreshUserInfo()).resolves.toMatchSnapshot();
        expect(result.current.loginUser).toMatchSnapshot();
      });

    });

    describe('logined', () => {

      beforeEach(async () => {
        authMock.verify.mockResolvedValueOnce(new VerifyResponse({
          userName: 'user1',
          expireAt: Timestamp.fromDate(new Date("2022/11/4")),
          requirePasswordUpdate: false,
        }));
        userMock.getUser.mockResolvedValue(new GetUserResponse({
          user: { name: 'user1', roles: ["CosmoAdmin"], displayName: 'user1 name' },
        }));
      });

      it('✅ ok', async () => {
        const { result } = await renderUseLogin();
        await act(async () => {
          await expect(result.current.refreshUserInfo()).resolves.toMatchSnapshot();
        });
        expect(result.current.loginUser).toMatchSnapshot();
      });

    });
  });


  describe('logout', () => {

    beforeEach(async () => {
      authMock.verify.mockResolvedValueOnce(new VerifyResponse({
        userName: 'user1',
        expireAt: Timestamp.fromDate(new Date("2022/11/4")),
        requirePasswordUpdate: false,
      }));
      userMock.getUser.mockResolvedValue(new GetUserResponse({
        user: { name: 'user1', roles: ["CosmoAdmin"], displayName: 'user1 name' },
      }));
    });

    it('✅ ok', async () => {
      const { result } = await renderUseLogin();
      await act(async () => {
        await expect(result.current.logout()).resolves.toMatchSnapshot();
      });
      await waitFor(async () => { expect(result.current.loginUser).toBeUndefined(); });
      expect(authMock.logout.mock.calls).toMatchSnapshot('logout calls');
    });

    it('❌ error', async () => {
      const { result } = await renderUseLogin();
      authMock.logout.mockRejectedValue(new Error('[mock] logout error'));
      await act(async () => {
        await expect(result.current.logout()).rejects.toMatchSnapshot();
      });
      await waitFor(async () => { expect(result.current.loginUser).toBeUndefined(); });
      expect(authMock.logout.mock.calls).toMatchSnapshot('logout calls');
    });

  });


  describe('updatePassword', () => {

    beforeEach(async () => {
      authMock.verify.mockResolvedValueOnce(new VerifyResponse({
        userName: 'user1',
        expireAt: Timestamp.fromDate(new Date("2022/11/4")),
        requirePasswordUpdate: false,
      }));
      userMock.getUser.mockResolvedValue(new GetUserResponse({
        user: { name: 'user1', roles: ["CosmoAdmin"], displayName: 'user1 name' },
      }));
    });

    it('✅ ok', async () => {
      const { result } = await renderUseLogin();
      userMock.updateUserPassword.mockResolvedValue(new UpdateUserPasswordResponse({
        message: "ok",
      }));
      await expect(result.current.updataPassword('oldpw', 'newpw')).resolves.toMatchSnapshot();
      expect(userMock.updateUserPassword.mock.calls).toMatchSnapshot('putUserPassword calls');
    });

    it('❌ error', async () => {
      const { result } = await renderUseLogin();
      userMock.updateUserPassword.mockRejectedValue(new Error('[mock] updateUserPassword error'));
      await expect(result.current.updataPassword('oldpw', 'newpw')).rejects.toMatchSnapshot();
      expect(userMock.updateUserPassword.mock.calls).toMatchSnapshot('putUserPassword calls');
    });

  });

});
