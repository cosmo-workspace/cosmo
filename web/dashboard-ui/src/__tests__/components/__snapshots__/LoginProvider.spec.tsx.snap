// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginProvider > login > ✅ ok 1`] = `
{
  "expireAt": "2022-11-03T15:00:00Z",
  "requirePasswordUpdate": true,
  "userName": "user1",
}
`;

exports[`LoginProvider > login > ✅ ok 2`] = `
{
  "addons": [],
  "authType": "",
  "defaultPassword": "",
  "displayName": "user1 name",
  "name": "user1",
  "roles": [
    "CosmoAdmin",
  ],
  "status": "",
}
`;

exports[`LoginProvider > login > ✅ ok 3`] = `[]`;

exports[`LoginProvider > login > ❌ getUser error 1`] = `[Error: [mock] getUser error]`;

exports[`LoginProvider > login > ❌ getUser error 2`] = `undefined`;

exports[`LoginProvider > login > ❌ getUser error 3`] = `
[
  [
    "[mock] getUser error",
    {
      "variant": "error",
    },
  ],
  [
    "[mock] getUser error",
    {
      "variant": "error",
    },
  ],
]
`;

exports[`LoginProvider > login > ❌ ng 1`] = `[Error: [mock] login error]`;

exports[`LoginProvider > login > ❌ ng 2`] = `undefined`;

exports[`LoginProvider > login > ❌ ng 3`] = `
[
  [
    "[mock] login error",
    {
      "variant": "error",
    },
  ],
]
`;

exports[`LoginProvider > logout > ✅ ok > logout calls 1`] = `
[
  [
    {},
  ],
]
`;

exports[`LoginProvider > logout > ✅ ok 1`] = `undefined`;

exports[`LoginProvider > logout > ✅ ok 2`] = `[]`;

exports[`LoginProvider > logout > ❌ error > logout calls 1`] = `
[
  [
    {},
  ],
]
`;

exports[`LoginProvider > logout > ❌ error 1`] = `[Error: [mock] logout error]`;

exports[`LoginProvider > logout > ❌ error 2`] = `
[
  [
    "[mock] logout error",
    {
      "variant": "error",
    },
  ],
]
`;

exports[`LoginProvider > refreshUserInfo > logined > ✅ ok 1`] = `undefined`;

exports[`LoginProvider > refreshUserInfo > logined > ✅ ok 2`] = `
{
  "addons": [],
  "authType": "",
  "defaultPassword": "",
  "displayName": "user1 name",
  "name": "user1",
  "roles": [
    "CosmoAdmin",
  ],
  "status": "",
}
`;

exports[`LoginProvider > refreshUserInfo > logined > ✅ ok 3`] = `[]`;

exports[`LoginProvider > refreshUserInfo > not login > ✅ ok 1`] = `undefined`;

exports[`LoginProvider > refreshUserInfo > not login > ✅ ok 2`] = `undefined`;

exports[`LoginProvider > refreshUserInfo > not login > ✅ ok 3`] = `[]`;

exports[`LoginProvider > updatePassword > ✅ ok > putUserPassword calls 1`] = `
[
  [
    {
      "currentPassword": "oldpw",
      "newPassword": "newpw",
      "userName": "user1",
    },
  ],
]
`;

exports[`LoginProvider > updatePassword > ✅ ok 1`] = `
{
  "message": "ok",
}
`;

exports[`LoginProvider > updatePassword > ✅ ok 2`] = `[]`;

exports[`LoginProvider > updatePassword > ❌ error > putUserPassword calls 1`] = `
[
  [
    {
      "currentPassword": "oldpw",
      "newPassword": "newpw",
      "userName": "user1",
    },
  ],
]
`;

exports[`LoginProvider > updatePassword > ❌ error 1`] = `[Error: [mock] updateUserPassword error]`;

exports[`LoginProvider > updatePassword > ❌ error 2`] = `
[
  [
    "[mock] updateUserPassword error",
    {
      "variant": "error",
    },
  ],
]
`;

exports[`LoginProvider > verify > ✅ ok 1`] = `
{
  "addons": [],
  "authType": "",
  "defaultPassword": "",
  "displayName": "user1 name",
  "name": "user1",
  "roles": [
    "CosmoAdmin",
  ],
  "status": "",
}
`;

exports[`LoginProvider > verify > ✅ ok 2`] = `[]`;

exports[`LoginProvider > verify > ❌ getUser error 1`] = `undefined`;

exports[`LoginProvider > verify > ❌ getUser error 2`] = `
[
  [
    "[mock] getUser error",
    {
      "variant": "error",
    },
  ],
]
`;

exports[`LoginProvider > verify > ❌ ng / id is undefined 1`] = `undefined`;

exports[`LoginProvider > verify > ❌ ng / id is undefined 2`] = `[]`;

exports[`LoginProvider > verify > ❌ ng 1`] = `undefined`;

exports[`LoginProvider > verify > ❌ ng 2`] = `[]`;
